barretenberg/cpp/src/barretenberg/stdlib/hash/poseidon2/poseidon2_permutation.hpp

```diff
@@ -0,0 +1,112 @@
+                                             current_state[2].witness_index,
+                                             current_state[3].witness_index,
+                                             i };
+            ctx->create_poseidon2_external_gate(in);
+            // calculate the new witnesses
+            NativePermutation::add_round_constants(current_native_state, round_constants[i]);
+            NativePermutation::apply_sbox(current_native_state);
+            NativePermutation::matrix_multiplication_external(current_native_state);
+            for (size_t j = 0; j < t; ++j) {
+                current_state[j] = field_t<Builder>(ctx, current_native_state[j]);
+            }
+        }
+
+        const size_t p_end = rounds_f_beginning + rounds_p;
+        for (size_t i = rounds_f_beginning; i < p_end; ++i) {
+            poseidon2_internal_gate_<FF> in{ current_state[0].witness_index,
+                                             current_state[1].witness_index,
+                                             current_state[2].witness_index,
+                                             current_state[3].witness_index,
+                                             i };
+            ctx->create_poseidon2_internal_gate(in);
+            current_native_state[0] += round_constants[i][0];
+            NativePermutation::apply_single_sbox(current_native_state[0]);
+            NativePermutation::matrix_multiplication_internal(current_native_state);
+            for (size_t j = 0; j < t; ++j) {
+                current_state[j] = field_t<Builder>(ctx, current_native_state[j]);
+            }
+        }
+
+        for (size_t i = p_end; i < NUM_ROUNDS; ++i) {
+            poseidon2_external_gate_<FF> in{ current_state[0].witness_index,
+                                             current_state[1].witness_index,
+                                             current_state[2].witness_index,
+                                             current_state[3].witness_index,
+                                             i };
+            ctx->create_poseidon2_external_gate(in);
+            // calculate the new witnesses
+            NativePermutation::add_round_constants(current_native_state, round_constants[i]);
+            NativePermutation::apply_sbox(current_native_state);
+            NativePermutation::matrix_multiplication_external(current_native_state);
+            for (size_t j = 0; j < t; ++j) {
+                current_state[j] = field_t<Builder>(ctx, current_native_state[j]);
+            }
+        }
+        // need to add an extra row here to ensure that things check out
+        poseidon2_end_gate_<FF> in{
+            current_state[0].witness_index,
+            current_state[1].witness_index,
+            current_state[2].witness_index,
+            current_state[3].witness_index,
+        };
+        ctx->create_poseidon2_end_gate(in);
+        return current_state;
+    }
+};
```

Comment:
could just create normal add gate here instead

Reply:
this got moved to a different line for some reason, but it was referring to the rounds which only use 3 of the wires and only do an add (gates 4 and 6)